s/types.go

```diff
@@ -0,0 +1,181 @@
+package s
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/capatazlib/go-capataz/c"
+)
+
+// Order specifies the order in which a supervision tree is going to start and
+// stop its children. The stop ordering is always the reverse of the start
+// ordering.
+type Order uint32
+
+const (
+	// LeftToRight is an Order that specifies children start from left to right
+	LeftToRight Order = iota
+	// RightToLeft is an Order that specifies children start from right to left
+	RightToLeft
+)
+
+// SortStart returns children sorted for the supervisor start
+func (o Order) SortStart(input0 []c.Spec) []c.Spec {
+	input := append(input0[:0:0], input0...)
+	switch o {
+	case LeftToRight:
+		return input
+	case RightToLeft:
+		for i, j := 0, len(input)-1; i < j; i, j = i+1, j-1 {
+			input[i], input[j] = input[j], input[i]
+		}
+		return input
+	default:
+		panic("Invalid s.Order value")
+	}
+}
+
+// SortStop returns children sorted for the supervisor stop
+func (o Order) SortStop(input0 []c.Spec) []c.Spec {
+	input := append(input0[:0:0], input0...)
+	switch o {
+	case LeftToRight:
+		for i, j := 0, len(input)-1; i < j; i, j = i+1, j-1 {
+			input[i], input[j] = input[j], input[i]
+		}
+		return input
+	case RightToLeft:
+		return input
+	default:
+		panic("Invalid s.Order value")
+	}
+}
+
+// Strategy specifies how children get restarted when one of them reports an
+// error
+type Strategy uint32
+
+const (
+	// OneForOne is an Strategy that tells the Supervisor to only restart the
+	// child process that errored
+	OneForOne Strategy = iota
+	// OneForAll
+	// RestForOne
+)
+
+// EventTag specifies the type of Event that gets notified from the supervision
+// system
+type EventTag uint32
+
+const (
+	eventNoop EventTag = iota
+	// ProcessStarted is an Event that indicates a process started
+	ProcessStarted
+	// ProcessStopped is an Event that indicates a process was stopped by a parent
+	// supervisor
+	ProcessStopped
+	// ProcessFailed is an Event that indicates a process reported an error
+	ProcessFailed
+)
+
+// String returns a string representation of the current EventTag
+func (tag EventTag) String() string {
+	switch tag {
+	case ProcessStarted:
+		return "ProcessStarted"
+	case ProcessStopped:
+		return "ProcessStopped"
+	case ProcessFailed:
+		return "ProcessFailed"
+	default:
+		return "<Unknown>"
+	}
+}
+
+// Event is a record emitted by the supervision system. The events are used for
+// multiple purposes, from testing to monitoring the healthiness of the
+// supervision system.
+type Event struct {
+	tag     EventTag
+	name    string
```

Comment:
Can you elaborate?

Reply:
I mean, do you really need the `name string` field given that you have an `EventTag.String()` function? Do you expect the possibility of an `Event.name` being different than the `EventTag` stringified version?

Reply:
Oh  I see, the name attribute here is the name of the Child/Supervisor that reported the event, maybe it would make sense to rename it to `processRuntimeName`. An `Event` info is more than that `String`, you may notice somewhere in the code we have a `String()` function already defined